Home Screen: Add Store Management (Quick Links) card

mattsherman opened the pull request:
Fixes woocommerce/woocommerce-admin#4082

This PR adds the new "Store Management" (Quick Links) card to the home screen.

### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

<img width="1263" alt="ScreenCapture at Sat May 16 06:44:50 EDT 2020" src="https://user-images.githubusercontent.com/2098816/82117964-ecd0d980-9741-11ea-92e7-97220e69a79d.png">

### Detailed test instructions

- Enable debugging of Tracks in your browser console:

```
localStorage.setItem( 'debug', 'wc-admin:*' );
```

- View the new home screen.
- Verify that the "Store Management" card is shown.
- Click items and verify that they navigate to the correct locations (from woocommerce/woocommerce-admin#4082) 
    - When the user clicks "Market my store" he will be redirected to WooCommerce - Marketing
    - When the user clicks "Add products" he will be redirected to WooCommerce - Products
    - When the user clicks "Personalize my store" he will be redirected to Appearance - Customize
    - When the user clicks "Shipping settings" he will be redirected to WooCommerce - Settings - Shipping
    - When the user clicks "Tax settings" he will be redirected to WooCommerce - Settings - Tax
    - When the user clicks "Payment settings" he will be redirected to WooCommerce - Settings - Payments
    - When the user clicks "Edit store details" he will be redirected to WooCommerce - Settings - General
    - When the user clicks "View my store" he will be redirected to the site frontend
- Verify that `wcadmin_home_quick_links_click` Tracks events are recorded, with the `task_name` custom prop set appropriately, every time an item in the list is clicked.

### Unit tests

```
npm test -- -i client/quick-links/test/index.js
```

### Known issues

- Icons do not match design. (/cc @jameskoster Where can I find the correct icons to use?)
- Card always shows on home screen. It should only appear when the "Set up your store and start selling" task list card has been completed or manually dismissed. This will be addressed after the task list card has been merged (#4378)
- Card has incorrect ARIA roles (#4387). This should be addressed separately, as it is a problem with the underlying `<List>` component.
- Support link always goes to premium support. This should be addressed separately, as it is a core problem with support links in general (https://github.com/woocommerce/woocommerce/issues/25442), and a solution should be implemented to make sure that support links are consistent throughout the product.
- Storybook does not show styles correctly for `<QuickLinks>` component

<!--- Note: When displaying information based on sample data, such as SwaggerHub, 
be sure to detail parts affected in Release Notes --->

### Changelog Note:

<!--- Optional: Enter a changelog note following the WooCommerce core format using prefixes of Enhancement:, Tweak:, Dev:, Fix:, Performance:. If no note is entered, one will be constructed from the title and labels. --->

Enhancement: New "Store Management" quick links card on WooCommerce home screen.

mattsherman committed:
Initial QuickLinks component commit.

mattsherman committed:
Add Quick Links card to home page

mattsherman committed:
Fix href creation to use relative urls

mattsherman committed:
Get `siteUrl` for link to view store.

mattsherman committed:
Convert QuickLinks to a functional component.

mattsherman committed:
Fix typo in unit test.

mattsherman committed:
Support link

mattsherman committed:
Personalize my store link

mattsherman committed:
Quick Links card header styling

mattsherman committed:
Update colors

mattsherman committed:
Define hover color in theme, as a workaround

mattsherman committed:
Move sass variables scope and use `color` (css) instead of `darken` (sass)

mattsherman committed:
Margin between homepage items

mattsherman committed:
Move "Store management" under "Stats overview".

mattsherman committed:
Merge branch 'master' into add/homepage-quick-links

psealock commented:
I went the items to test and just noting it down here while I continue the review.

 - [x] Verify that the "Store Management" card is shown.
 - [x] When the user clicks "Market my store" he will be redirected to WooCommerce - Marketing
 - [x] When the user clicks "Add products" he will be redirected to WooCommerce - Products
 - [x] When the user clicks "Personalize my store" he will be redirected to Appearance - Customize
 - [x] When the user clicks "Shipping settings" he will be redirected to WooCommerce - Settings - Shipping
 - [x] When the user clicks "Tax settings" he will be redirected to WooCommerce - Settings - Tax
 - [x] When the user clicks "Payment settings" he will be redirected to WooCommerce - Settings - Payments
 - [x] When the user clicks "Edit store details" he will be redirected to WooCommerce - Settings - General
 - [x] When the user clicks "View my store" he will be redirected to the site frontend
 - [x] Verify that wcadmin_home_quick_links_click Tracks events are recorded, with the task_name custom prop set appropriately, every time an item in the list is clicked.

psealock commented on client/homepage/style.scss:
```diff
@@ -13,18 +13,41 @@
+
+		& > * {
+			margin-bottom: 24px;
+		}
```
👍

psealock commented on client/homepage/style.scss:
```diff
@@ -13,18 +13,41 @@
 	}
+
+	.components-card {
+		.components-card__header.is-size-large {
```
Are these rules going to apply to all large cards on the home page or just this one? If just this one, maybe something specific to the quicklinks would be best.

mattsherman replied:
I'd expect we'd want to apply it to all large cards on the home page, for consistency.

mattsherman replied:
Further, I'd expect that all cards on the home page should use the `@wordpress/components` `Card`.

psealock replied:
Cool, I saw that issue 👍

psealock commented on client/homepage/style.scss:
```diff
@@ -13,18 +13,41 @@
+			display: grid;
+			//align-items: center;
+			min-height: 63px;
+			padding: 13px 16px;
```
We should stick with `$gap` and `$gap-small` if possible.

jameskoster replied:
Is it possible to use the [`Card`](https://wordpress.github.io/gutenberg/?path=/story/components-card--default) component here? Then we don't have to worry about things like padding values.

mattsherman replied:
I'm already using the `@wordpress/components` `Card` component. But, the resulting alignment without tweaking things, differs from both the `@woocommerce/components` `Card` component and the design Figma.

I've updated the code to use the gap variables, as suggested.

psealock commented on client/homepage/style.scss:
```diff
@@ -13,18 +13,41 @@
+	.components-card {
+		.components-card__header.is-size-large {
+			display: grid;
+			//align-items: center;
```
errant commented out code

mattsherman replied:
Oops. Removed!

psealock commented on client/quick-links/index.js:
```diff
@@ -0,0 +1,190 @@
+	];
+}
+
+function onItemClick( props, event ) {
```
How about `handleOnItemClick` since `onItemClick` is already a prop?

mattsherman replied:
Will update.

psealock commented on client/quick-links/index.js:
```diff
@@ -0,0 +1,190 @@
+		icon = <Icon icon={ iconName } />;
+	}
+
+	return icon;
```
Looks like a really handy util function. I wonder if we could use this elsewhere?

mattsherman replied:
Agreed. But, depending on what ends up happening in regards to what icons are used, we might no need this function for this component. If we end up keeping it, I can refactor it out to a common component/function.

psealock commented on client/quick-links/index.js:
```diff
@@ -0,0 +1,190 @@
+			after: getIcon( 'arrow-right-alt2' ),
+			...getLinkTypeAndHref( item ),
+			listItemTag: item.listItemTag,
+			onClick: partial( onItemClick, props ),
```
Nice, just noting here that we did some investigation already to determine if this function gets called before page navigation and in particular if Tracks events used in this way do indeed get recorded. It works as it should 😄

psealock commented on client/quick-links/index.js:
```diff
@@ -0,0 +1,190 @@
+	return (
+		<Card size="large" className="woocommerce-quick-links">
+			<CardHeader>
+				<h2>{ __( 'Store management', 'woocommerce-admin' ) }</h2>
```
There is a handy container for heading levels to automatically frame out accessible heading hierarchy, `<H />`.

https://github.com/woocommerce/woocommerce-admin/blob/577d362f5149f86f68500a8fdd71ed4a9149065a/packages/components/src/section/index.js#L22-L31

mattsherman replied:
Will use, thanks for letting me know!

psealock commented on client/quick-links/style.scss:
```diff
@@ -0,0 +1,70 @@
+			margin: 0;
+			padding: 0;
+
+			.woocommerce-list__item {
```
These selectors are 4 and 5 levels deep, any way to make it less?

mattsherman replied:
I've improved this a little bit.

psealock commented on client/quick-links/style.scss:
```diff
@@ -0,0 +1,70 @@
+				// and updating that causes compatibility issues with the version of
+				// `@wordpress/components` that we are using, and updating that introduces
+				// breaking changes that we would need to address elsewhere in the
+				// `woocommerce-admin` codebase.
```
Great documentation! Should we add to it by creating an issue and linking it here?

mattsherman replied:
I actually created an issue last week. I'll add a link to it in the code comment!

psealock reviewed (commented):
This is looking great @mattsherman and testing well ! Just some comments about a few relatively minor things.

jameskoster reviewed (dismissed):
@mattsherman the icons are a part of the [WordPress core icons package](https://github.com/WordPress/gutenberg/tree/master/packages/icons) and should be accessible via the `Icon` component. :)

mattsherman committed:
Updating homepage styles based on feedback.

mattsherman committed:
Update item onClick handler function name to avoid confusion with handler prop.

mattsherman committed:
Add link to issue about updating `@wordpress/base-styles` and `@wordpress/components`.

mattsherman committed:
Use `<H>` component instead of `<h2>`.

mattsherman commented:
@jameskoster All of the icons used in the design (https://www.figma.com/file/Z2PiNnkhjrUIiR7IrfJRnU/Home-Screen?node-id=970%3A0) are not available in the `Icon` component. And, those that do differ in style from what you used.

It would be very helpful if you could provide me with the name of each icon to use, so that I can ensure that things are implemented as desired. Right now, I'm using a combination of `Icon` (and therefore `dashicon`) and `gridicons` icons to pick things that look close to what the design showed.

mattsherman committed:
Reduce nesting in styles.

mattsherman committed:
Merge branch 'add/homepage-quick-links' of github.com:woocommerce/woocommerce-admin into add/homepage-quick-links

mattsherman commented:
@psealock Thanks for the review and feedback! I've addressed your feedback in the latest commits.

If you feel the PR is ready to be approved outside of the icon changes, I can always take care of those in a followup PR.

psealock reviewed (approved):
This is looking fantastic and thanks for the changes. LGTM and good to follow up with icons separately.

jameskoster commented:
> All of the icons used in the design are not available in the Icon component. And, those that do differ in style from what you used.

Hm, is that the core Icon component? They should be there because I [added](https://github.com/WordPress/gutenberg/pull/21209) [them](https://github.com/WordPress/gutenberg/pull/21825). 

At least, I added the ones that I thought were missing, maybe some others need updating from their Dashicon counterparts 🤔 

Here's the explicit list of icons to use;

- Market my store - [Megaphone](https://github.com/WordPress/gutenberg/blob/master/packages/icons/src/library/megaphone.js)
- Add products - [Box](https://github.com/WordPress/gutenberg/blob/master/packages/icons/src/library/box.js)
- Personalize my store - [Brush](https://github.com/WordPress/gutenberg/blob/master/packages/icons/src/library/brush.js)
- Shipping settings - [Shipping](https://github.com/WordPress/gutenberg/blob/master/packages/icons/src/library/shipping.js)
- Tax settings - [Percent](https://github.com/WordPress/gutenberg/blob/master/packages/icons/src/library/percent.js)
- Payment settings - [Payment](https://github.com/WordPress/gutenberg/blob/master/packages/icons/src/library/payment.js)
- Edit store details - [Pencil](https://github.com/WordPress/gutenberg/blob/master/packages/icons/src/library/pencil.js)
- Get support - [Lifesaver](https://github.com/WordPress/gutenberg/blob/master/packages/icons/src/library/lifesaver.js)
- View my store - [External](https://github.com/WordPress/gutenberg/blob/master/packages/icons/src/library/external.js)

It looks like the Brush and Pencil icons will need to be updated to match the visual design of the others, but we'll handle that in core :)

Sorry for the confusion here.

mattsherman commented:
> > All of the icons used in the design are not available in the Icon component. And, those that do differ in style from what you used.
> 
> Hm, is that the core Icon component? They should be there because I [added](https://github.com/WordPress/gutenberg/pull/21209) [them](https://github.com/WordPress/gutenberg/pull/21825).
> 
> At least, I added the ones that I thought were missing, maybe some others need updating from their Dashicon counterparts 🤔
> 
> Here's the explicit list of icons to use;
> [...]
> Sorry for the confusion here.

@jameskoster Thanks for the links to the icons to use. I think the confusion stemmed from my use of `Icon` from `@wordpress/components` instead of from `@wordpress/icons`. I've created a new issue to address updating the icons (#4396), so that I can merge this PR. I'll get a PR up for that and make sure to request a review from you on it so we can ensure that the correct icons are used!
